In [360]:
import pandas as pd
import getpass
import sqlalchemy as sa
import numpy as np
import seaborn as sns
import matplotlib as plt
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import quantile_transform
from sklearn.preprocessing import QuantileTransformer
from sklearn.preprocessing import minmax_scale
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder  ##. better to use dummy from pandas 
from sklearn.preprocessing import PowerTransformer
from scipy.stats import boxcox
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
from scipy.stats import boxcox
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae
from sklearn.feature_selection import RFE
pd.options.display.max_rows = 50
## Install xlrd package to load Excel files
# conda install openpyxl
## conda install xlrd

## Importing Dataset.

In [361]:
hp_df_org = pd.read_csv('C:/Users/Lenovo/Documents/GitHub/Houseprices/Data/house_price_df.csv')
hp_df = hp_df_org

## EDA

In [362]:
hp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     21597 non-null  int64  
 1   id             21597 non-null  int64  
 2   date           21597 non-null  object 
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     21597 non-null  int64  
 9   view           21597 non-null  int64  
 10  condition      21597 non-null  int64  
 11  grade          21597 non-null  int64  
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  int64  
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   21597 non-null  int64  
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

The Dataset consists of 21597 Houses containing 21 features. Price is to be the predicted variables. 

In [363]:
hp_df.describe()

Unnamed: 0            id      bedrooms     bathrooms   sqft_living  \
count  21597.000000  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   10798.000000  4.580474e+09      3.373200      2.115826   2080.321850   
std     6234.661218  2.876736e+09      0.926299      0.768984    918.106125   
min        0.000000  1.000102e+06      1.000000      0.500000    370.000000   
25%     5399.000000  2.123049e+09      3.000000      1.750000   1430.000000   
50%    10798.000000  3.904930e+09      3.000000      2.250000   1910.000000   
75%    16197.000000  7.308900e+09      4.000000      2.500000   2550.000000   
max    21596.000000  9.900000e+09     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  21597.000000   
mean   1.509941e+04      1.494096      0.007547      0.234292      3.409825   
std    4.141264e+04      0.539683      0.086549      0.766390      0.650546   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068500e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

       ...    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  ...  21597.000000   21597.000000  21597.000000  21597.000000   
mean   ...   1788.596842     291.725008   1970.999676     84.464787   
std    ...    827.759761     442.667800     29.375234    401.821438   
min    ...    370.000000       0.000000   1900.000000      0.000000   
25%    ...   1190.000000       0.000000   1951.000000      0.000000   
50%    ...   1560.000000       0.000000   1975.000000      0.000000   
75%    ...   2210.000000     560.000000   1997.000000      0.000000   
max    ...   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  \
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000   
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512   
std       53.513072      0.138552      0.140724     685.230472   27274.441950   
min    98001.000000     47.155900   -122.519000     399.000000     651.000000   
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000   
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000   
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000   
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000   

              price  
count  2.159700e+04  
mean   5.402966e+05  
std    3.673681e+05  
min    7.800000e+04  
25%    3.220000e+05  
50%    4.500000e+05  
75%    6.450000e+05  
max    7.700000e+06  

[8 rows x 21 columns]

The average house has about 3 bedrooms, 2080 sqft of living area, a lot size of 15000 sqft, was built in 1970 and has a price of 540000 $. The mean and the median differ alot in the variable sqft lot, where the median is much lower then the mean. This leads to the conculusion that many houses don't have a big lot but a few house with a big lot change the distribution.

In [ ]:
sns.pairplot(hp_df)

In [ ]:
print(hp_df['date'].min(), hp_df['date'].max())

The date variable displays the date when the house was sold. Data is between May 2014 and May 2015. 

In [ ]:
hp_df['bathrooms'].value_counts()

Most houses have 2,5 bathrooms, a lot also only . A view outliers have a very high number of bathrooms.

## Location

For the location the datasets offers three features. Latitute and Longitude (only useful combined) and zipcodes. Let's look at both of them.

In [ ]:
 sns.scatterplot(data=hp_df, x='long', y='lat', hue = 'waterfront').set(title='Houses at waterfront')

Looking at the geographical location of the houses and displaying the lot size it is visual, that the properties at the are outskirts and the countryside get bigger. 

In [ ]:
 sns.scatterplot(data=hp_df, x='long', y='lat', hue = 'zipcode').set(title='Zipcode and geographical location')

The zipcodes are not in a logical order so the variables should not be used as a numerical variables.
And alternative idea would be to use the travel distance from the house to the city center.

In [ ]:
distance = pd.read_csv('C:/Users/Lenovo/Documents/GitHub/Houseprices/Data/distance.csv')

In [ ]:
hp_df['distanceM'] = distance['DistanceM'].astype('float64')

In [ ]:
sns.scatterplot(data=hp_df, x='long', y='lat', hue = 'distanceM').set(title='Geographical location and distance to citycenter by car')

## Year renovated

In [ ]:
sns.histplot(data=hp_df, x='yr_renovated')

The year renovated displays a 0 if the house was not yet renovated. This is not useful information in a linear regression. 

Workaround:

introduce boolean variable: Renovated: Yes/N0
Put zero values as No  


In [ ]:
hp_df['renovated'] = bool

In [ ]:
hp_df['yr_renovated'] == 0

In [ ]:
hp_df.loc[hp_df['yr_renovated'] == 0, 'renovated'] = False

In [ ]:
hp_df.loc[hp_df['yr_renovated'] != 0, 'renovated'] = True

## Basement

In [ ]:
sns.histplot(data=hp_df, x='sqft_basement')

Most Houses don't have a basement, so a dummy will be introduced checking for basement or not:

In [ ]:
hp_df.loc[hp_df['sqft_basement'] == 0, 'basement'] = False
hp_df.loc[hp_df['sqft_basement'] != 0, 'basement'] = True

## Grade

Bin grade to have a better classification:

In [ ]:
grade_mean = hp_df['grade'].mean()

In [ ]:
for i in range(0,len(hp_df)):
    if hp_df['grade'][i]<grade_mean:
        hp_df['grade'][i] = 'below_avg'
    else:
        hp_df['grade'][i] = 'avg_or_better'
        

In [ ]:
sns.histplot(hp_df['grade'])

## Floors

In [ ]:
sns.histplot(hp_df['floors'])

In [ ]:
hp_df['floors'].value_counts()

bin floors to be more equal

In [ ]:
for i in range(0, len(hp_df)):
    if hp_df.loc[i, 'floors'] <= 1.0:
        hp_df.loc[i,'floors'] = 'one'
    else: 
        hp_df.loc[i, 'floors'] = 'more_than_one'

In [ ]:
sns.histplot(hp_df['floors'])

## Dropping columns:

For now the columns Unnamed, id, date and long, lat will be dropped from the dataframe.

In [ ]:
col_drop = ['Unnamed: 0',\
            'id',\
            'date',\
           'long',\
            'sqft_basement',\
            'yr_renovated',\
           'lat']   
# 'Unnamed: 0' : id from import without information
# 'id': random or consequtive values without values
# 'date': probably the date the house was added to the database, no additional information for houseprice
# 'lang', 'lat': Geografical data, not usable in linear regression like this, can be converted to zones using knn. 
#    Task for later. But zip codes are enough, probably


In [ ]:
hp_df = hp_df.drop(col_drop, axis= 1)

## Changing datatypes


The format of the features are mostly numerical (int or float). Many of the variables have to be transformed to category to give meaningful results in a regression.
These are:

'zipcode, 'bedrooms', 'bathrooms', 'floors', 'waterfront', 'view', 'condition', 'grade'

In [ ]:
hp_df.info()

In [ ]:
# From the pairplot we can see which columns are categorical/dummies:
#  bderooms, bathrooms, floors, waterfront, view, condition, grade
ordinal_var = ['zipcode', 'floors', 'waterfront', 'view', 'condition', 'grade']
hp_df[ordinal_var] = hp_df[ordinal_var].astype('category')
hp_df.info()

In [ ]:
hp_df.shape

## Checking correlation of numerical variables

In [ ]:
corr = hp_df.corr()
mask = np.zeros_like(corr)
mask[np.triu_indices_from(mask)] = True
with sns.axes_style("white"):
    f, ax = plt.subplots(figsize=(9, 7))
    ax = sns.heatmap(corr, mask=mask,cmap='coolwarm', vmin=-1,vmax=1,annot=True, square=True)

The features should not be correlated in order to fulfill the asumtions of linear regression.
As expected there is a lot of correlation between sqf_above, sqf_living and other measures of size. 
- First: Drop sqf_above

In [ ]:
corr_matrix=hp_df.corr().abs()
upper_triangle=corr_matrix.where(np.triu(np.ones(corr_matrix.shape),k=1).astype(bool))
corr_var = [column for column in upper_triangle.columns if any(upper_triangle[column] > 0.80)]
hp_df = hp_df.drop(corr_var, axis= 1)
hp_df.shape

In [ ]:
corr = hp_df.corr()
mask = np.zeros_like(corr)
mask[np.triu_indices_from(mask)] = True
with sns.axes_style("white"):
    f, ax = plt.subplots(figsize=(9, 7))
    ax = sns.heatmap(corr, mask=mask,cmap='coolwarm', vmin=-1,vmax=1,annot=True, square=True)

The new dataframe does not contain anymore variables with a higher correlation than 0.8, but some variable still have high correlation which could be analysed later.

# Normalization

## Histograms
To check for the assumtion of normal distributions we check the histograms of the numerical variables

In [ ]:
hp_df.select_dtypes(['int64','float']).hist(figsize=(12,12), bins=40)


All numerical varibales are non-normaly distributed and need to be transformed. 

## Transforming to normal distribution

In [ ]:
hp_df.select_dtypes(['int64','float']).columns

In [ ]:
# List of variables names to be transformed: 
to_trans =  {'sqft_living': 'qt',\
             'sqft_lot': 'qt',\
             'yr_built': 'qt',\
             'sqft_living15': 'qt',\
             'sqft_lot15': 'qt',\
            'bedrooms': 'qt',\
            'bathrooms': 'qt',\
            'distanceM': 'qt'}

For the transformation of the variables the Quantile Transformer is applied. 

In [ ]:

pt = PowerTransformer()
qt = QuantileTransformer(output_distribution="normal")

transformed_cols = []
for i in to_trans.keys():
    if list(to_trans.values())[0] == 'pt':
        transformed = pt.fit_transform(hp_df[i].to_numpy().reshape(-1,1))
    elif list(to_trans.values())[0] == 'qt':
        transformed = qt.fit_transform(hp_df[i].to_numpy().reshape(-1,1))
    else: 
        print('no transformer could be identified')
    col_name = i+'_transformed'
    sns.displot(transformed).set(title=f'Histogram of {col_name}')
    hp_df[col_name] = transformed
    transformed_cols = transformed_cols + [col_name]
    hp_df = hp_df.drop(i, axis= 1)
    
    
hp_df.describe()

In [ ]:
transformed_cols

In [ ]:
hp_df.select_dtypes(['int', 'float64']).hist(figsize=(12,12), bins=40)

## Remove outliers

The Datasets contains a few datapoints that can described as outliers. One house for example as 33 rooms while the average amount of rooms is around 3. Using the two time the Interquartile range whe remove the outliers. 

In [ ]:
for i in transformed_cols:
    sns.boxplot(data=hp_df[transformed_cols]).set(title='Boxplots of numerical variables showing outliers')
    plt.xticks(rotation=90)

In [ ]:
 len_before = hp_df.shape[0]

In [ ]:
def remove_outlier(df_in, col_name):
    q1 = df_in[col_name].quantile(0.25)
    q3 = df_in[col_name].quantile(0.75)
    iqr = q3-q1 #Interquartile range
    fence_low  = q1-2*iqr
    fence_high = q3+2*iqr
    df_out = df_in.loc[(df_in[col_name] > fence_low) & (df_in[col_name] < fence_high)]
    return df_out

#for i in transformed_cols: 
hp_df = remove_outlier(hp_df, i )

In [ ]:
for i in transformed_cols:
    sns.boxplot(data=hp_df[transformed_cols]).set(title='Boxplots of numerical variables showing outliers')
    plt.xticks(rotation=90)

In [ ]:
len_after = hp_df.shape[0]

In [ ]:
print(f'Of the original dataset {round(len_after/len_before * 100, 2)}% of rows remain')

## One-hot encode categorical variables

In [ ]:
col_names_object = list(hp_df.select_dtypes("object").columns)
col_names_object.extend(hp_df.select_dtypes("category").columns)
col_names_object

In [ ]:
cat_features =  pd.get_dummies(hp_df[col_names_object] , drop_first=True)

In [ ]:
hp_df = hp_df.drop(col_names_object, axis = 1)

In [ ]:
hp_df[cat_features.columns] = cat_features

In [ ]:
list(hp_df.columns)

## 1st Model only using numerical variables

###  1) Split X-y

In [ ]:
X=hp_df.drop('price', axis=1)
y=hp_df['price']


In [ ]:
X

In [ ]:
X=X.drop(cat_features.columns, axis=1)

In [ ]:
X.columns

In [ ]:
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=.30,random_state=100)

In [ ]:
X_train

### Standardize

In [ ]:
std_scaler=StandardScaler().fit(X_train)   ##. finding the parameters ( mean, variance from the training set )
                                            ## 
X_train_scaled=std_scaler.transform(X_train)

In [ ]:
X_train.shape

In [ ]:
X_test_scaled=std_scaler.transform(X_test)
X_test_scaled.shape

## Linear Regression

In [ ]:
X_train_const_scaled = sm.add_constant(X_train_scaled) # adding a constant

model = sm.OLS(y_train, X_train_const_scaled).fit()
predictions_train = model.predict(X_train_const_scaled) 

X_test_const_scaled = sm.add_constant(X_test_scaled) # adding a constant
predictions_test = model.predict(X_test_const_scaled) 
print_model = model.summary()
print(print_model)

In [ ]:
model=LinearRegression()    # model
model.fit(X_train_scaled, y_train)   # model train

In [ ]:
model.coef_

In [ ]:
y_pred=model.predict(X_test_scaled)   # model prediction

y_pred_train=model.predict(X_train_scaled)

In [ ]:
result=pd.DataFrame({"y_test":y_test,"y_pred":y_pred})

In [ ]:
# Make an scatter plot y_pred vs y
# What kind of plot you will get if all the all the predictions are ok?
# A stright line

fig, ax = plt.subplots(1,3,figsize=(14,4))
ax[0].plot(y_pred, y_test, 'o')
ax[0].set_xlabel("y_test")
ax[0].set_ylabel("y_pred")
ax[0].set_title("Test Set -Predicted vs real")

# Get a histogram of the residuals ie: y - y_pred.  Homoscdasticity
# It resembles a normal distribution?
ax[1].hist(y_test - y_pred)
ax[1].set_xlabel("Test y-y_pred")
ax[1].set_title("Test Set Residual histogram")

ax[2].plot(y_pred,y_test - y_pred,"o")
ax[2].set_xlabel("predicted")
ax[2].set_ylabel("residuals")
ax[2].set_title("Residuals by Predicted")
ax[2].plot(y_pred,np.zeros(len(y_pred)),linestyle='dashed')

In [ ]:
print('MSE test: ',  mse(y_test,y_pred))
print('MAE test: ', mae(y_test,y_pred))
print('MSE train: ', mse(y_train,y_pred_train))
print('RMSE test: ', np.sqrt(mae(y_test,y_pred)))
R2_test=model.score(X_test_scaled,y_test)
R2_train=model.score(X_train_scaled,y_train)
print('R2_test: ', R2_test)
print('R2_train: ', R2_train)

print('MSE train - MSE test: ', mse(y_train,y_pred_train) - mse(y_test,y_pred) )
print('R2_test/R2_train: ', R2_test/R2_train )
Adj_R2= 1 - (1-R2_test)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)
print('Adj_R2:', Adj_R2)

In [ ]:
features_importances = pd.DataFrame(data={
    'Attribute': X_train.columns,
    'Importance': abs(model.coef_)
})
features_importances = features_importances.sort_values(by='Importance', ascending=False)

In [ ]:
features_importances

In [ ]:
plt.bar(x=features_importances['Attribute'], height=features_importances['Importance'], color='#087E8B')
plt.title('Feature importances obtained from coefficients', size=20)
plt.xticks(rotation='vertical')
plt.show()

## 2nd model including categorical variables

In [ ]:
def Xy_split(df, output_col):
    """ Split the dataset into:
        X = dataframe of input variables
        y = predicted variable
    """
    X=df.drop(output_col, axis=1)
    y=df[output_col]
    print(X.columns)
    return X, y

X,y = Xy_split(hp_df, 'price') 

In [ ]:
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=.30,random_state=100)

In [ ]:
std_scaler=StandardScaler().fit(X_train.select_dtypes(['int', 'float64']))   ##. finding the parameters ( mean, variance from the training set )
                                            ## 
X_train_scaled=std_scaler.transform(X_train.select_dtypes(['int', 'float64']))

X_test_scaled=std_scaler.transform(X_test.select_dtypes(['int', 'float64']))
X_test_scaled.shape

In [ ]:
X_train

In [ ]:
X_train_scaled = np.concatenate((X_train_scaled, np.asarray(X_train[list(cat_features.columns)])), axis = 1) 

X_test_scaled = np.concatenate((X_test_scaled, np.asarray(X_test[list(cat_features.columns)])), axis =1)
print(X_train_scaled.shape, X_test_scaled.shape)

In [ ]:
print(y_train.shape, y_test.shape)

In [ ]:
model=LinearRegression()    # model
model.fit(X_train_scaled, y_train) 


y_pred=model.predict(X_test_scaled)   # model prediction
y_pred_train=model.predict(X_train_scaled)


result=pd.DataFrame({"y_test":y_test,"y_pred":y_pred})


# Make an scatter plot y_pred vs y
# What kind of plot you will get if all the all the predictions are ok?
# A stright line

fig, ax = plt.subplots(1,3,figsize=(14,4))
ax[0].plot(y_pred, y_test, 'o')
ax[0].set_xlabel("y_test")
ax[0].set_ylabel("y_pred")
ax[0].set_title("Test Set -Predicted vs real")

# Get a histogram of the residuals ie: y - y_pred.  Homoscdasticity
# It resembles a normal distribution?
ax[1].hist(y_test - y_pred)
ax[1].set_xlabel("Test y-y_pred")
ax[1].set_title("Test Set Residual histogram")

ax[2].plot(y_pred,y_test - y_pred,"o")
ax[2].set_xlabel("predicted")
ax[2].set_ylabel("residuals")
ax[2].set_title("Residuals by Predicted")
ax[2].plot(y_pred,np.zeros(len(y_pred)),linestyle='dashed')

In [ ]:
y_test

In [ ]:
#sns.jointplot(y_pred, y_test)

In [ ]:
#sns.jointplot(x=y_pred,y = (y_test - y_pred))

In [ ]:
print('MSE test: ',  mse(y_test,y_pred))
print('MAE test: ', mae(y_test,y_pred))
print('MSE train: ', mse(y_train,y_pred_train))
print('RMSE test: ', np.sqrt(mae(y_test,y_pred)))
R2_test=model.score(X_test_scaled,y_test)
R2_train=model.score(X_train_scaled,y_train)
print('R2_test: ', R2_test)
print('R2_train: ', R2_train)

print('MSE train - MSE test: ', mse(y_train,y_pred_train) - mse(y_test,y_pred) )
print('R2_test/R2_train: ', R2_test/R2_train )
Adj_R2= 1 - (1-R2_test)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)
print('Adj_R2:', Adj_R2)

In [ ]:
features_importances = pd.DataFrame(data={
    'Attribute': X_train.columns,
    'Importance': abs(model.coef_)
})
features_importances = features_importances.sort_values(by='Importance', ascending=False)
features_importances
plt.bar(x=features_importances['Attribute'], height=features_importances['Importance'], color='#087E8B')
plt.title('Feature importances obtained from coefficients', size=20)
plt.xticks(rotation='vertical')
plt.figure(figsize=(200, 200), dpi=300)
plt.show()

In [ ]:
features_importances.nlargest(20, 'Importance')

## Get P-values from model 

To filter the significant columns we get the p-value from the linear regression. If it is higher than 0.5 the column is dropped.

In [ ]:
X_train_const_scaled = sm.add_constant(X_train_scaled) # adding a constant

model = sm.OLS(y_train, X_train_const_scaled).fit()
predictions_train = model.predict(X_train_const_scaled) 

X_test_const_scaled = sm.add_constant(X_test_scaled) # adding a constant
predictions_test = model.predict(X_test_const_scaled) 
print_model = model.summary()
print(print_model)

In [ ]:
sign_cols = model.pvalues[1:len(model.pvalues)]  # remove constant

sign_cols

Pseudo-Code: for i in range(0, len(X.column)) drop X.columns[i] if sign_cols[i]> 0.05
        

In [ ]:
sign_cols = pd.DataFrame([sign_cols])
sign_cols

In [ ]:
X_col_list = X.columns
X_dropped_p = X
for i in range(0,len(X.columns)):
    if sign_cols.iloc[0,i]>0.05:
        X_dropped_p = X_dropped_p.drop(X_col_list[i], axis=1)

In [ ]:
print(f'From orginally {X.shape[1]} features {X_dropped_p.shape[1]} features remain.')

## 3d linear Model only including variabels that had a siginificant p-value in the 2s Model

In [ ]:
X_train, X_test, y_train, y_test=train_test_split(X_dropped_p, y, test_size=.30,random_state=100)

In [ ]:
X_train_num = X_train.select_dtypes(['float64'])
X_test_num = X_test.select_dtypes(['float64'])
X_train_cat = X_train.select_dtypes(['uint8'])
X_test_cat = X_test.select_dtypes(['uint8'])
print(X_train_num.shape, X_test_num.shape, X_train_cat.shape, X_test_cat.shape)

In [ ]:
X_train.info()

In [ ]:
std_scaler=StandardScaler().fit(X_train_num)   ##. finding the parameters ( mean, variance from the training set )
                                            ## 
X_train_scaled=std_scaler.transform(X_train_num)

X_test_scaled=std_scaler.transform(X_test_num)
X_test_scaled.shape

In [ ]:
X_train_scaled = np.concatenate((X_train_scaled, np.asarray(X_train_cat)), axis = 1) 

X_test_scaled = np.concatenate((X_test_scaled, np.asarray(X_test_cat)), axis =1)
print(X_train_scaled.shape, X_test_scaled.shape)

In [ ]:
model=LinearRegression()    # model
model.fit(X_train_scaled, y_train) 


y_pred=model.predict(X_test_scaled)   # model prediction
y_pred_train=model.predict(X_train_scaled)


result=pd.DataFrame({"y_test":y_test,"y_pred":y_pred})


# Make an scatter plot y_pred vs y
# What kind of plot you will get if all the all the predictions are ok?
# A stright line

fig, ax = plt.subplots(1,3,figsize=(14,4))
ax[0].plot(y_pred, y_test, 'o')
ax[0].set_xlabel("y_test")
ax[0].set_ylabel("y_pred")
ax[0].set_title("Test Set -Predicted vs real")

# Get a histogram of the residuals ie: y - y_pred.  Homoscdasticity
# It resembles a normal distribution?
ax[1].hist(y_test - y_pred)
ax[1].set_xlabel("Test y-y_pred")
ax[1].set_title("Test Set Residual histogram")

ax[2].plot(y_pred,y_test - y_pred,"o")
ax[2].set_xlabel("predicted")
ax[2].set_ylabel("residuals")
ax[2].set_title("Residuals by Predicted")
ax[2].plot(y_pred,np.zeros(len(y_pred)),linestyle='dashed')

In [ ]:
print('MSE test: ',  mse(y_test,y_pred))
print('MAE test: ', mae(y_test,y_pred))
print('MSE train: ', mse(y_train,y_pred_train))
print('RMSE test: ', np.sqrt(mae(y_test,y_pred)))
R2_test=model.score(X_test_scaled,y_test)
R2_train=model.score(X_train_scaled,y_train)
print('R2_test: ', R2_test)
print('R2_train: ', R2_train)

print('MSE train - MSE test: ', mse(y_train,y_pred_train) - mse(y_test,y_pred) )
print('R2_test/R2_train: ', R2_test/R2_train )
Adj_R2= 1 - (1-R2_test)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)
print('Adj_R2:', Adj_R2)

In [ ]:
features_importances = pd.DataFrame(data={
    'Attribute': X_train.columns,
    'Importance': abs(model.coef_)
})
features_importances = features_importances.sort_values(by='Importance', ascending=False)
features_importances
plt.bar(x=features_importances['Attribute'], height=features_importances['Importance'], color='#087E8B')
plt.title('Feature importances obtained from coefficients', size=20)
plt.xticks(rotation='vertical')
plt.show()

In [ ]:
features_importances #.nlargest(30, 'Importance')

In [ ]:
features_importances.to_csv('C:/Users/Lenovo/Documents/GitHub/Houseprices/Data/final_model_feature_importance.csv')

In [ ]:
X_train_const_scaled = sm.add_constant(X_train_scaled) # adding a constant

model = sm.OLS(y_train, X_train_const_scaled).fit()
predictions_train = model.predict(X_train_const_scaled) 

X_test_const_scaled = sm.add_constant(X_test_scaled) # adding a constant
predictions_test = model.predict(X_test_const_scaled) 
print_model = model.summary()
print(print_model)

## Visualizing results:

Zip-code showed to be very determining for prices. The zipcode 98039 showed to show the highest price effect.

In [ ]:
avg_grouped_by_zipcode = hp_df_org.groupby('zipcode').mean()

In [ ]:
avg_grouped_by_zipcode.drop(['Unnamed: 0', 'id'], axis=1, inplace=True)

In [ ]:
avg_grouped_by_zipcode.columns

In [ ]:
sns.pairplot(avg_grouped_by_zipcode)

In [ ]:
sns.scatterplot(data=avg_grouped_by_zipcode, x='long', y='lat', hue='bathrooms')

In [ ]:
sns.scatterplot(data=avg_grouped_by_zipcode, x='long', y='lat', hue='zipcode')

Filtering for postcode with biggest influence: 98039

In [ ]:
zipcode98039 = hp_df_org[hp_df_org['zipcode']==98039]

In [ ]:
sns.scatterplot(data=zipcode98039, x='long', y='lat')

In [ ]:
zipcode98004 = hp_df_org[hp_df_org['zipcode']==98004]

In [ ]:
waterfront1 = hp_df_org[hp_df_org['waterfront']==1]
waterfront1.describe()

In [ ]:
sns.scatterplot(data=waterfront1, x='long', y='lat')

Get Map background:

In [ ]:
import geopandas
import contextily as cx

In [ ]:
def vplot(df, xax_var, yax_var, split_var):
    '''
    Function to create a violin plot with given variables
    xax_var = 'sales_channel' # Categorical variable to show along the x-axis
    yax_var = 'total_claim_amount' # Numerical Variable to show along the y-axis
    split_var = 'response' #Boolean variable to display on each half of the violins
    '''
    return sns.violinplot(x =df[xax_var],\
                        y = yax_var,\
                        hue=split_var,\
                        data= hp_df_org, split=True)

In [ ]:
xax_var = 'view' # Categorical variable to show along the x-axis
yax_var = 'price' # Numerical Variable to show along the y-axis
split_var = 'waterfront' #Boolean variable to display on each half of the violins
vplot(hp_df_org, xax_var, yax_var, split_var)

In [ ]:
xax_var = 'waterfront' # Categorical variable to show along the x-axis
yax_var = 'price' # Numerical Variable to show along the y-axis
split_var = 'view' #Boolean variable to display on each half of the violins
vplot(hp_df_org, xax_var, yax_var, split_var)